src/com/hms/flexyosisoftconnector/OsisoftJsonPayload.java

```diff
@@ -0,0 +1,382 @@
+                dataPoint.getValueString(), timestamp, isFirstPoint);
+        break;
+      case OsisoftConfig.piwebapi:
+        newPayloadPart = PayloadBuilder.addPointOldFormat(dataPoint, dataPointsEncounteredCount);
+        break;
+      default:
+        Logger.LOG_SERIOUS(comsErrMsg);
+        newPayloadPart = "no communication type";
+        break;
+    }
+
+    ((TagPayload) tagPayloadArr[index]).add(newPayloadPart);
+
+    // might remove all this if statement
+    if (status == PAYLOAD_HEADER_COMPLETE) {
+
+      status = PAYLOAD_VALUES_IN_PROGRESS;
+
+      long time = Long.parseLong(dataPoint.getTimeStamp());
+      payloadStartTimestamp = time;
+    }
+
+    // add to the count of data points
+    dataPointsEncounteredCount++;
+  }
+
+  private void didReachMaxDataPoints() {
+
+    if (dataPointsEncounteredCount >= MAX_DATA_POINTS) {
+      status = PAYLOAD_VALUES_COMPLETE;
+    }
+  }
+
+  /**
+   * Convert payload start timestamp from epoch seconds to ISO 8601 date format in UTC.
+   *
+   * @return formatted payload start timestamp
+   */
+  private String getFormattedStartTime(long time) {
+    // move to constants somewhere
+    int TIME_MILLIS_PER_SEC = 1000;
+    Date d = new Date(time * TIME_MILLIS_PER_SEC);
+    SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-ddTHH:mm:ss'Z'");
+    dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+    String timestamp = dateFormat.format(d);
+    return timestamp;
+  }
+
+  /**
+   * Inspect a tag ID. If it has been seen, start a new payload, otherwise, add to the tags
+   * encountered array.
+   *
+   * @param tagId ID of tag to inspect
+   */
+  private int tagIdToIndex(int tagId) {
+    int index = tagId - tagIndexOffset;
+    return index;
+  }
+
+  /**
+   * Completes the payload if it has values
+   *
+   * @return true if payload was completed
+   */
+  public void completePayloadAttempt() {
+    if (status == PAYLOAD_VALUES_COMPLETE) {
+      // Payload has values, end the payload
+      String batchTimeStamp = getFormattedStartTime(payloadStartTimestamp);
+
+      boolean firstPayload = true;
+
+      // add all the tag's info to the payload
+      for (int i = 0; i < tagPayloadArr.length; i++) {
+
+        if (tagPayloadArr[i] != null && tagPayloadArr[i].payload.length() > 1) {
+
+          TagPayload tagPayload = (TagPayload) tagPayloadArr[i];
+
+          String tagName = tagPayload.dataPoint.getTagName();
+
+          if (!firstPayload) {
+            payload.append(",");
+          } else {
+            firstPayload = false;
+          }
+
+          payload.append(PayloadBuilder.addContainerStartToOMFDataMessage(tagName));
+
+          payload.append(tagPayload.payload);
+
+          payload.append(PayloadBuilder.addContainerEndToOMFDataMessage());
+        }
+      }
+
+      boolean allAppended = payload.append(PayloadBuilder.endOMFDataMessage());
+
+      if (!allAppended) {
+        Logger.LOG_SERIOUS(
```

Comment:
sounds a bit confusing. Maybe something like "An error occurred while appending a payload to a message. Datapoints may be lost."